### ISSUE TITLE
[FR] Allow docker args to be passed into `mlflow models build-docker` and `mlflow sagemaker build-and-push-container` 

### ISSUE THREAD

**Issue by tobiasbartsch** [mlflow/mlflow#11484]
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

During the build-docker process, mlflow installs packages as specified in a requirements.txt file that is logged together with the model in the mlflow registry.
Some of these packages may be in a private package repository, whose access requires a token. There is currently no way to pass such a token into the build process.

### Motivation

> #### What is the use case for this feature?
During the build-docker process, mlflow installs packages as specified in a requirements.txt file that is logged together with the model in the mlflow registry.
Some of these packages may be in a private package repository, whose access requires a token. There is currently no way to pass such a token (as an ARG) into the build process.

> #### Why is this use case valuable to support for MLflow users in general?
Currently the documentation recommends to add code not installable from pipy as `code` to the log_model call. This does not scale well since adding packages as `code` does not resolve their dependencies.

> #### Why is this use case valuable to support for your project(s) or organization?
See above.

> #### Why is it currently difficult to achieve this use case?
See above.

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [X] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by WeichenXu123** [comment]
Can we specify the docker image environment variable instead ?

**Comment by tobiasbartsch** [comment]
Unfortunately not -- if we'd pass such a token in as an ENV then it would persist in the image (and anyone with access to the image could retrieve the token, which is a security risk). However, if we pass the token as an ARG, it is only present while the image is building.

**Comment by tobiasbartsch** [comment]
That said, it might be nice to support both. There are probably use cases to set environment variables at the time `mlflow models build-docker` is called.

**Comment by WeichenXu123** [comment]
The feature request makes sense.

**Comment by ctaymor** [comment]
Hi, I'd be interested in helping with this.

However, I have some security concerns about the proposed approach. [Docker does not recommend using ARGs for secrets.](https://docs.docker.com/build/building/variables/) 
There are three ways secrets passed as ARGs can be exposed:
* **Docker History**: Secrets passed as ARGs are recorded and visible via `docker history`
* **Layer Persistence**: Due to Docker's layered architecture, secrets persist in image layers even if deleted
  later
* **Metadata Exposure**: Build arguments become part of the image metadata

The more secure approach would be to use [Docker Secret Mounts](https://docs.docker.com/build/building/secrets/). We could conditionally pass the secret to PIP when the secret is provided.

However, Docker Secret Mounts are only available when running with Buildkit. If a user is using an older version of Docker without Buildkit (which became the default in Docker Engine 23.0 in March 2023) a secret mount won't work.

To ensure backwards compatibility, we could use a multi stage build approach, where the secret is used in the first stage, and the docker image is finalized in a second stage. This is not as secure, because the secrets will still be in the build cache on the machine building the image. However, with this approach, they are at least not distributed with the image.

Many enterprise and CI environments do run older versions of Docker without Buildkit, so the backwards compatibility is likely important.

# Alternative approaches:
1. Multi-stage build + conditional build secrets OR ARGs
We could switch to a multi-stage build for all images, and when there are private repositories in the requirements.txt, either use secrets or ARGs. (We can't directly detect Buildkit from within Docker, but there are some proxies we can use to detect it, such as checking the docker version, or trying the secrets mount and falling back to an ARG if it is not available.)

2. Multi-stage build with ARGs
We could add ARGs for this secret, and use a multi-stage build. If we assume few users are using buildkit, this is the most backwards compatible option. However, then no users can take advantage of the increased security of secret mounts.

3. Allow the user to specify whether they are using buildkit, or which security method to use
This might be easier and more reliable than detecting whether buildkit is used, however, it would require a change in workflow. (Then again, if you're providing a secret, then that is a change in workflow also).
For instance, if you provide a secret, then you also need to provide`--auth-method=buildkit|multistage`.

4. Only support buildkit for private repos and use secret mounts
Non-buildkit would still be supported when a secret is not provided.

This would be the most secure, technically, however, my guess is it would leave at least some of our users behind. I don't know for sure, though. It might be that this is a reasonable constraint.

# My Thoughts
@WeichenXu123 @tobiasbartsch I'm leaning toward option 1, as I think it provides the best tradeoff between backwards compatibility and security, however, I'd love to hear your thoughts on the tradeoffs.

**Comment by endesapt** [comment]
@ctaymor I agree with you that option 1 is the best one, but i think that the flows of creating of building image in case of multi-stage and secret mount is too different. Implementing this might be a problem.
I suppose we need to unite 1 and 2 points together: we need to use multistage build everytime, but inside the first stage:
1. Use secret mount if possible(docker version is high enough)
2. use args otherwise

As i understand the first stage will download dependencies , and second one will copy those dependencies and run the model as earlier
I can try to implement it if @WeichenXu123 @tobiasbartsch think it is still necessary
